nixos/modules/security/apparmor.nix

```
@@ -90,13 +125,17 @@ in
   };
 
   config = lib.mkIf cfg.enable {
-    assertions = map (policy:
-      { assertion = match ".*/.*" policy == null;
-        message = "`security.apparmor.policies.\"${policy}\"' must not contain a slash.";
+    assertions =
+      map (policyName: {
+        assertion = match ".*/.*" policyName == null;
+        message = "`security.apparmor.policies.\"${policyName}\"' must not contain a slash.";
         # Because, for instance, aa-remove-unknown uses profiles_names_list() in rc.apparmor.functions
         # which does not recurse into sub-directories.
-      }
-    ) (attrNames cfg.policies);
+      }) (attrNames cfg.policies)
+      ++ lib.mapAttrsToList (policyName: policyCfg: {
+        assertion = lib.xor (policyCfg.path != null) (submoduleOptionIsDefined policyCfg.profile);
+        message = "`security.apparmor.policies.\"${policyName}\"` must define exactly one of either path or profile.";
+      }) cfg.policies;
```

```suggestion
      lib.concatLists (
        lib.mapAttrsToList (policyName: policyCfg: [
          {
            # Because, for instance, aa-remove-unknown uses profiles_names_list() in rc.apparmor.functions
            # which does not recurse into sub-directories.
            assertion = match ".*/.*" policyName == null;
            message = "`security.apparmor.policies.\"${policyName}\"' must not contain a slash.";
          }
          {
            assertion = lib.xor (policyCfg.path != null) (submoduleOptionIsDefined policyCfg.profile);
            message = "`security.apparmor.policies.\"${policyName}\"` must define exactly one of either path or profile.";
          }
        }) cfg.policies
      );
```

Since we're mapping over the same value, it'd be better to define all of the assertions here rather than separately